In [5]:
!kaggle datasets download -d debasisdotcom/parkinson-disease-detection

from zipfile import ZipFile
import os
for i in os.listdir('C:/Users/mehmu/Desktop/PyProjects/Projects/github repo/ParkinsonDetection/DataScience'):
    if '.zip' in i:
        print(i)
        with ZipFile(i, 'r') as zipObj:
            zipObj.extractall()
        os.remove(i)

parkinson-disease-detection.zip: Skipping, found more recently modified local copy (use --force to force download)
parkinson-disease-detection.zip


In [1]:
import pandas as pd
import numpy as np
import seaborn as sbn
import matplotlib.pyplot as plt

df=pd.read_csv('Parkinsson disease.csv')
df

name  MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
0    phon_R01_S01_1      119.992       157.302        74.997         0.00784   
1    phon_R01_S01_2      122.400       148.650       113.819         0.00968   
2    phon_R01_S01_3      116.682       131.111       111.555         0.01050   
3    phon_R01_S01_4      116.676       137.871       111.366         0.00997   
4    phon_R01_S01_5      116.014       141.781       110.655         0.01284   
..              ...          ...           ...           ...             ...   
190  phon_R01_S50_2      174.188       230.978        94.261         0.00459   
191  phon_R01_S50_3      209.516       253.017        89.488         0.00564   
192  phon_R01_S50_4      174.688       240.005        74.287         0.01360   
193  phon_R01_S50_5      198.764       396.961        74.904         0.00740   
194  phon_R01_S50_6      214.289       260.277        77.973         0.00567   

     MDVP:Jitter(Abs)  MDVP:RAP  MDVP:PPQ  Jitter:DDP  MDVP:Shimmer  ...  \
0             0.00007   0.00370   0.00554     0.01109       0.04374  ...   
1             0.00008   0.00465   0.00696     0.01394       0.06134  ...   
2             0.00009   0.00544   0.00781     0.01633       0.05233  ...   
3             0.00009   0.00502   0.00698     0.01505       0.05492  ...   
4             0.00011   0.00655   0.00908     0.01966       0.06425  ...   
..                ...       ...       ...         ...           ...  ...   
190           0.00003   0.00263   0.00259     0.00790       0.04087  ...   
191           0.00003   0.00331   0.00292     0.00994       0.02751  ...   
192           0.00008   0.00624   0.00564     0.01873       0.02308  ...   
193           0.00004   0.00370   0.00390     0.01109       0.02296  ...   
194           0.00003   0.00295   0.00317     0.00885       0.01884  ...   

     Shimmer:DDA      NHR     HNR  status      RPDE       DFA   spread1  \
0        0.06545  0.02211  21.033       1  0.414783  0.815285 -4.813031   
1        0.09403  0.01929  19.085       1  0.458359  0.819521 -4.075192   
2        0.08270  0.01309  20.651       1  0.429895  0.825288 -4.443179   
3        0.08771  0.01353  20.644       1  0.434969  0.819235 -4.117501   
4        0.10470  0.01767  19.649       1  0.417356  0.823484 -3.747787   
..           ...      ...     ...     ...       ...       ...       ...   
190      0.07008  0.02764  19.517       0  0.448439  0.657899 -6.538586   
191      0.04812  0.01810  19.147       0  0.431674  0.683244 -6.195325   
192      0.03804  0.10715  17.883       0  0.407567  0.655683 -6.787197   
193      0.03794  0.07223  19.020       0  0.451221  0.643956 -6.744577   
194      0.03078  0.04398  21.209       0  0.462803  0.664357 -5.724056   

      spread2        D2       PPE  
0    0.266482  2.301442  0.284654  
1    0.335590  2.486855  0.368674  
2    0.311173  2.342259  0.332634  
3    0.334147  2.405554  0.368975  
4    0.234513  2.332180  0.410335  
..        ...       ...       ...  
190  0.121952  2.657476  0.133050  
191  0.129303  2.784312  0.168895  
192  0.158453  2.679772  0.131728  
193  0.207454  2.138608  0.123306  
194  0.190667  2.555477  0.148569  

[195 rows x 24 columns]

In [2]:
df.columns

Index(['name', 'MDVP:Fo(Hz)', 'MDVP:Fhi(Hz)', 'MDVP:Flo(Hz)', 'MDVP:Jitter(%)',
       'MDVP:Jitter(Abs)', 'MDVP:RAP', 'MDVP:PPQ', 'Jitter:DDP',
       'MDVP:Shimmer', 'MDVP:Shimmer(dB)', 'Shimmer:APQ3', 'Shimmer:APQ5',
       'MDVP:APQ', 'Shimmer:DDA', 'NHR', 'HNR', 'status', 'RPDE', 'DFA',
       'spread1', 'spread2', 'D2', 'PPE'],
      dtype='object')

In [3]:
df.drop('name',axis=1,inplace=True)

In [4]:
X_train = df.drop("status", axis=1)
y_train = df.status

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

In [5]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [6]:
from sklearn.model_selection import cross_validate, GridSearchCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.linear_model import SGDClassifier
from sklearn.naive_bayes import GaussianNB
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier,StackingClassifier


from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import classification_report, accuracy_score, f1_score, confusion_matrix

In [7]:
classifiers = {
    'Catboost': CatBoostClassifier(),
    "NB": GaussianNB(),
    "SGD": SGDClassifier(),
    "KNN": KNeighborsClassifier(), 
    "LR": LogisticRegression(max_iter=1000), 
    "DT": DecisionTreeClassifier(),
    "RF": RandomForestClassifier(n_estimators=100, max_depth=3, random_state=2),
    "SVM": SVC(),
    "MLP": MLPClassifier(max_iter=1000),
    "XGB": XGBClassifier(),
    "LGBM": LGBMClassifier(),
    'GradB':GradientBoostingClassifier()
}


results = pd.DataFrame(columns=["Classifier", "Avg_Accuracy", "Avg_F1_Score"])
for name, clf in classifiers.items():
    model = clf
    cv_results = cross_validate(
        model, X_test_scaled, y_test, cv=10,
        scoring=(['accuracy', 'f1'])
    )

    results = results.append({
        "Classifier": name,
        "Avg_Accuracy": cv_results['test_accuracy'].mean(),
        "Avg_F1_Score": cv_results['test_f1'].mean()
    }, ignore_index=True)
    
results["Avg_Overall"] = (results["Avg_Accuracy"] + results["Avg_F1_Score"]) / 2
results = results.sort_values("Avg_Overall", ascending=False)
results

c:\Users\mehmu\Desktop\PyProjects\env\lib\site-packages\sklearn\model_selection\_split.py:680: UserWarning: The least populated class in y has only 7 members, which is less than n_splits=10.
  warnings.warn(


Learning rate set to 0.002462
0:	learn: 0.6911011	total: 135ms	remaining: 2m 14s
1:	learn: 0.6892051	total: 135ms	remaining: 1m 7s
2:	learn: 0.6873514	total: 136ms	remaining: 45.2s
3:	learn: 0.6853796	total: 137ms	remaining: 34.1s
4:	learn: 0.6829942	total: 137ms	remaining: 27.3s
5:	learn: 0.6814606	total: 138ms	remaining: 22.9s
6:	learn: 0.6795631	total: 139ms	remaining: 19.7s
7:	learn: 0.6775276	total: 139ms	remaining: 17.3s
8:	learn: 0.6755775	total: 140ms	remaining: 15.4s
9:	learn: 0.6737712	total: 141ms	remaining: 13.9s
10:	learn: 0.6719959	total: 141ms	remaining: 12.7s
11:	learn: 0.6700083	total: 142ms	remaining: 11.7s
12:	learn: 0.6670288	total: 142ms	remaining: 10.8s
13:	learn: 0.6652307	total: 143ms	remaining: 10.1s
14:	learn: 0.6634691	total: 144ms	remaining: 9.42s
15:	learn: 0.6608847	total: 144ms	remaining: 8.86s
16:	learn: 0.6585726	total: 145ms	remaining: 8.36s
17:	learn: 0.6564200	total: 145ms	remaining: 7.92s
18:	learn: 0.6547507	total: 146ms	remaining: 7.53s
19:	learn:

C:\Users\mehmu\AppData\Local\Temp\ipykernel_36296\918575696.py:25: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results = results.append({
c:\Users\mehmu\Desktop\PyProjects\env\lib\site-packages\sklearn\model_selection\_split.py:680: UserWarning: The least populated class in y has only 7 members, which is less than n_splits=10.
  warnings.warn(
C:\Users\mehmu\AppData\Local\Temp\ipykernel_36296\918575696.py:25: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results = results.append({
c:\Users\mehmu\Desktop\PyProjects\env\lib\site-packages\sklearn\model_selection\_split.py:680: UserWarning: The least populated class in y has only 7 members, which is less than n_splits=10.
  warnings.warn(
C:\Users\mehmu\AppData\Local\Temp\ipykernel_36296\918575696.py:25: FutureWarning: The frame.append method is deprecated and wi

Classifier  Avg_Accuracy  Avg_F1_Score  Avg_Overall
11      GradB         0.900      0.938095     0.919048
4          LR         0.875      0.923810     0.899405
6          RF         0.875      0.923810     0.899405
9         XGB         0.875      0.923810     0.899405
0    Catboost         0.850      0.909524     0.879762
7         SVM         0.825      0.900000     0.862500
10       LGBM         0.825      0.900000     0.862500
3         KNN         0.825      0.895238     0.860119
8         MLP         0.825      0.895238     0.860119
2         SGD         0.825      0.883810     0.854405
5          DT         0.800      0.871429     0.835714
1          NB         0.800      0.863810     0.831905

In [11]:
svm = GradientBoostingClassifier()
params = {
    'learning_rate':[i*0.001 for i in range(1,101,5)],
    'max_depth':[i for i in range(1,10)],
    "n_estimators": [i for i in range(0,200,50)]
}
clf = GridSearchCV(svm, params, cv=10)
clf.fit(X_train_scaled, y_train)
print("Best hyperparameter:", clf.best_params_)

Best hyperparameter: {'learning_rate': 0.026000000000000002, 'max_depth': 3, 'n_estimators': 150}


c:\Users\mehmu\Desktop\PyProjects\env\lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
1800 fits failed out of a total of 7200.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1800 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\mehmu\Desktop\PyProjects\env\lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\mehmu\Desktop\PyProjects\env\lib\site-packages\sklearn\ensemble\_gb.py", line 577, in fit
    self._check_params()
  File "c:\Users\mehmu\Desktop\PyProjects\env\lib\site-packages\sklearn\ensemble\_gb.py", line 278, in _check_params
    check_scalar(
  File "c:\

In [12]:
results=[]
cv_results = cross_validate(
        svm, X_test_scaled, y_test, cv=10,
        scoring=(['accuracy', 'f1'])
    )

pd.DataFrame([cv_results['test_accuracy'].mean(),cv_results['test_f1'].mean()])

c:\Users\mehmu\Desktop\PyProjects\env\lib\site-packages\sklearn\model_selection\_split.py:680: UserWarning: The least populated class in y has only 7 members, which is less than n_splits=10.
  warnings.warn(


0
0  0.875000
1  0.918095

In [13]:
def scores(y_test,y_pred):
    from sklearn.metrics import confusion_matrix,precision_score,recall_score,f1_score
    cm=confusion_matrix(y_test,y_pred)
    ps=precision_score(y_test,y_pred),
    rs=recall_score(y_test,y_pred)
    f1=f1_score(y_test,y_pred)
    print(cm)
    df=pd.DataFrame([['precision_score',ps[0]],
    ['recall_score',rs],['f1',f1]],columns=['Metrics','Scores'])
    return df


y_pred=clf.predict(X_test_scaled)
scores(y_test,y_pred)


[[ 5  2]
 [ 1 31]]


Metrics    Scores
0  precision_score  0.939394
1     recall_score  0.968750
2               f1  0.953846

#### LGBM

[[ 5  2]
 [ 0 32]]
Metrics	Scores
0	precision_score	0.941176
1	recall_score	1.000000
2	f1	0.969697

#### GradientBoosting

[[ 5  2]
 [ 1 31]]
Metrics	Scores
0	precision_score	0.939394
1	recall_score	0.968750
2	f1	0.953846